graphql.go

```diff
@@ -0,0 +1,148 @@
+package goshopify
+
+import (
+	"math"
+	"time"
+)
+
+// GraphQLService is an interface to interact with the graphql endpoint
+// of the Shopify API
+// See https://shopify.dev/docs/admin-api/graphql/reference
+type GraphQLService interface {
+	Query(string, interface{}, interface{}) error
+}
+
+// GraphQLServiceOp handles communication with the graphql endpoint of
+// the Shopify API.
+type GraphQLServiceOp struct {
+	client *Client
+}
+
+type graphQLResponse struct {
+	Data       interface{}        `json:"data"`
+	Errors     []graphQLError     `json:"errors"`
+	Extensions *graphQLExtensions `json:"extensions"`
+}
+
+type graphQLExtensions struct {
+	Cost GraphQLCost `json:"cost"`
+}
+
+// GraphQLCost represents the cost of the graphql query
+type GraphQLCost struct {
+	RequestedQueryCost int                   `json:"requestedQueryCost"`
+	ActualQueryCost    *int                  `json:"actualQueryCost"`
+	ThrottleStatus     GraphQLThrottleStatus `json:"throttleStatus"`
+}
+
+// GraphQLThrottleStatus represents the status of the shop's rate limit points
+type GraphQLThrottleStatus struct {
+	MaximumAvailable   float64 `json:"maximumAvailable"`
+	CurrentlyAvailable float64 `json:"currentlyAvailable"`
+	RestoreRate        float64 `json:"restoreRate"`
+}
+
+type graphQLError struct {
+	Message    string                  `json:"message"`
+	Extensions *graphQLErrorExtensions `json:"extensions"`
+	Locations  []graphQLErrorLocation  `json:"locations"`
+}
+
+type graphQLErrorExtensions struct {
+	Code          string
+	Documentation string
+}
+
+const (
+	graphQLErrorCodeThrottled = "THROTTLED"
+)
+
+type graphQLErrorLocation struct {
+	Line   int `json:"line"`
+	Column int `json:"column"`
+}
+
+// Query creates a graphql query against the Shopify API
+// the "data" portion of the response is unmarshalled into resp
+func (s *GraphQLServiceOp) Query(q string, vars, resp interface{}) error {
+	data := struct {
+		Query     string      `json:"query"`
+		Variables interface{} `json:"variables"`
+	}{
+		Query:     q,
+		Variables: vars,
+	}
+
+	attempts := 0
+
+	for {
+		gr := graphQLResponse{
+			Data: resp,
+		}
+
+		err := s.client.Post("graphql.json", data, &gr)
+		// internal attempts count towards outer total
+		attempts += s.client.attempts
```

Comment:
I looked at the code again and I think this is a problem.
This is not threadsafe when multiple goroutines use the same client and make requests at the same time.
The state `attempts` part of the client is shared across requests and can't be relied on for truthfully telling the number of attempts made.
This should be something simple to show in a test (have the mock responder wait a sec for each request and respond with "throttled", then kick off 10 requests, all with the same client, then check `attempts`, should be the combined number of requests or so).

I think I'd want to remove `attempts` altogether and for instance have an alternate/secondary version of Post that returns the number of attempts.

This might be getting a bit much for this PR, to simplify you could, as first step, replace the `attempts+=...` with `attempts +=  1` andthen we can figure out later if this needs improving.

An alternate option would be to make the retry in `doGetHeaders()`  smarter and possibly make it aware of GraphQL. But again, that's a bigger effort and I'd be fine with the simplified `+=1`

Reply:
I went with the `+= 1` because I'm trying to ship something that depends on this, but I'll try to loop back and do a more proper management in another PR